### ISSUE TITLE
Some players seem to turn invisible, although they can still shoot

### ISSUE THREAD

**Issue by robhawkes** [robhawkes/rawkets#2]
At points I've noticed bullets being fired from invisible players. The only time I've noticed this has been when the game code has been manipulated by other players to bypass fire rates.

**Comment by robhawkes** [comment]
I've now witnessed this when the game isn't being manipulated by other players. The actual cause is yet unknown, but the main culprits are the methods that remove a player from the client code – causing it to not be drawn on the canvas any more.

**Comment by markembling** [comment]
Are you sure this isn't a deliberate manipulation?

**Comment by robhawkes** [comment]
Nope, but I've no way of knowing at the moment unless they own up to it.

**Comment by markembling** [comment]
I saw this earlier (pain in the bum because you can't shoot what you can't see, no matter how souped-up your gun is), and I assumed it was deliberate. Haven't seen it any other time, and when I was messing with mine I was never invisible (and hopefully wasn't for anyone else either). At worst, I'd tend to say that it was a 'bug' - and I use the term loosely - brought about by overzealous manipulation of the client-side JS.

**Comment by robhawkes** [comment]
Although, saying that I'm sure I'll be able to narrow it down to the specific API call that's causing the issues, wether it's manipulated or not.

**Comment by robhawkes** [comment]
Just thinking out aloud here, but it could be an issue with respawning in an illegal location. It might be worth me constructing some sort of debug system with which I can view player data.

**Comment by markembling** [comment]
Maybe. I've never respawned in an illegal location though (don't you have checks for that?), and I assumed that was dealt with server-side, not that I've checked.

**Comment by robhawkes** [comment]
Yeah, it shouldn't do it, but it's done client-side at the moment. I haven't really looked into this issue properly yet.

**Comment by markembling** [comment]
Ah right, well there you go then. In order to become invincible, you have to remove the behaviour you've provided for dealing with being killed. If you fudge that, I guess its possible that such an event could happen.

**Comment by robhawkes** [comment]
Do you mean invisible?

**Comment by markembling** [comment]
No, I mean invulnerable. I'm trying to say that invisibility nay be a side-effect of invulnerability hacks due to the need to muck around with the killing logic.

**Comment by robhawkes** [comment]
Ah yes, that could be possible. I'll update here once I've looked into it further.

**Comment by robhawkes** [comment]
Invisible players still seem to occur, but only when there is a lot of people in the game. Perhaps there is some bad logic that fails to update a kill/revival properly if it's overrun.

**Comment by robhawkes** [comment]
An interesting observation, although perhaps unrelated, is that there are some seemingly timed-out players with the flames still on. Could these be the bodies of the invisible players, or simply just timed-out players?

[Edit] Every time I see an invisible player now there seems to be a disembodied rawkets sitting stationery somewhere. Also, I haven't noticed an invisible player being able to kill anyone yet.

**Comment by robhawkes** [comment]
This bug actually happened to me today, although I'm unsure exactly what happened. I had left the game running in a tab while doing other word, but when I returned I noticed that other players couldn't see or shoot me. I could still shoot them.

**Comment by gavtaylor** [comment]
was happening to me last night, thought I was alone, but then bullets appeared out of nowhere, spent 5 minutes trying to avoid whoever it was but in the end got bored...
